internal/resources/package/data_source_package.go

```diff
@@ -40,3 +54,116 @@ var packageSchema = map[string]*schema.Schema{
 	common.MetaKey:       common.Meta,
 	spec.SpecKey:         spec.SpecSchema,
 }
+
+func resourcePackageRead(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+	config := m.(authctx.TanzuContext)
+
+	packageName, ok := d.Get(nameKey).(string)
+	if !ok {
+		return diag.Errorf("unable to read package name")
+	}
+
+	packageNamespaceName, ok := d.Get(namespaceKey).(string)
+	if !ok {
+		return diag.Errorf("Unable to read package namespace name")
+	}
+
+	metadataName, ok := d.Get(metadataNameKey).(string)
+	if !ok {
+		return diag.Errorf("Unable to read package metadata name")
+	}
+
+	scopedFullnameData := scope.ConstructScope(d, packageName, metadataName, packageNamespaceName)
+
+	if scopedFullnameData == nil {
+		return diag.Errorf("Unable to get Tanzu Mission Control package entry; Scope full name is empty")
+	}
+
+	UID, meta, atomicSpec, err := retrieveSourcesecretUIDMetaAndSpecFromServer(config, scopedFullnameData, d)
+	if err != nil {
+		if clienterrors.IsNotFoundError(err) {
+			if ctx.Value(contextMethodKey{}) == DataSourceRead {
+				return diag.FromErr(errors.Wrapf(err, "Tanzu Mission Control package entry not found, name : %s", d.Get(nameKey)))
+			}
+
+			_ = schema.RemoveFromState(d, m)
+
+			return diags
+		}
+
+		return diag.FromErr(err)
+	}
+
+	// always run
+	d.SetId(UID)
+
+	if err := d.Set(common.MetaKey, common.FlattenMeta(meta)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	flattenedSpec := spec.FlattenSpecForClusterScope(atomicSpec)
+
+	if err := d.Set(spec.SpecKey, flattenedSpec); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+func retrieveSourcesecretUIDMetaAndSpecFromServer(config authctx.TanzuContext, scopedFullnameData *scope.ScopedFullname, d *schema.ResourceData) (
+	string,
+	*objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta,
+	*tanzupackagemodel.VmwareTanzuManageV1alpha1ClusterNamespaceTanzupackageMetadataPackageSpec,
+	error) {
+	var (
+		UID  string
+		meta *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta
+		spec *tanzupackagemodel.VmwareTanzuManageV1alpha1ClusterNamespaceTanzupackageMetadataPackageSpec
+	)
+
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
+		if scopedFullnameData.FullnameCluster != nil {
+			resp, err := config.TMCConnection.TanzupackageResourceService.ManageV1alpha1ClusterPackageResourceServiceGet(scopedFullnameData.FullnameCluster)
+			if err != nil {
+				if clienterrors.IsNotFoundError(err) {
+					d.SetId("")
```

Comment:
Since you set the id **d.SetId("")** , what would happen to L89 which is executed after this?

Reply:
well, it doesn't return any error to set ID nil again but I've removed it from one place.

I've updated it PTAL.